enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,359 @@
+---
+title: spot-instances
+authors:
+  - "@JoelSpeed"
+  - "@enxebre"
+reviewers:
+  - "@enxebre"
+  - "@bison"
+  - "@michaelgugino"
+approvers:
+  - "@enxebre"
+  - "@bison"
+  - "@michaelgugino"
+creation-date: 2020-02-04
+last-updated: 2020-02-04
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Spot Instances
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Enable OCP users to leverage cheaper, non-guaranteed instances to back Machine API Machines.
+
+## Motivation
+
+Allow users to cut costs of running OCP clusters on cloud providers by moving interruptible workloads onto non-guaranteed instances.
+
+### Goals
+
+- Provide sophisticated provider-specific automation for running Machines on non-guaranteed instances
+
+- Utilise as much of the existing Machine API as possible
+
+- Ensure graceful shutdown of pods is attempted on non-guaranteed instances
+
+### Non-Goals
+
+- Any logic for choosing instances types based on availability from the cloud provider
+- A one to one map for each provider available mechanism for deploying spot instances, e.g aws fleet.
+
+## Proposal
+
+To provide a consistent behaviour using non-guaranteed instances (Spot on AWS and Azure, Preepmtible on GCP)
+across cloud providers, we must define a common behaviour based on the common features across each provider.
+
+Based on the research on [non-guaranteed instances](#non-guaranteed-instances),
+the following requirements for integration will work for each of AWS, Azure and GCP:
+
+- Required configuration for enabling spot/preemptible instances should be added to the ProviderSpec
```

Comment:
this is a good reason for us to revisit machine class so we can have a common definition so we can place these specifications in a common place and reference.

Reply:
I don't know what Machine classes are, is there anything I can read on these so I can work out how this would help here?

Reply:
Machine Classes are a way to externalise and decouple the providerSpec definition from the machine object in favour of an object reference. We should re-evaluate this in the near future in hand with the upstream provider templates https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/api/v1alpha3/awsmachinetemplate_types.go.
We can do this though orthogonally to this proposal.